tests/python_package_test/test_basic.py

```diff
@@ -913,6 +913,17 @@ def test_feature_names_are_set_correctly_when_no_feature_names_passed_into_Datas
     assert ds.construct().feature_name == ["Column_0", "Column_1", "Column_2"]
 
 
+@pytest.mark.parametrize("max_depth", [4, 8, 12, 16])
+def test_max_depth_is_enforced(capsys, max_depth):
+    X, y = make_blobs(n_samples=1_000, n_features=1, centers=2)
+    model = lgb.LGBMRegressor(objective="binary", max_depth=max_depth, device=getenv("TASK"), verbose=0)
+    model.fit(X, y)
+    assert (
+        model.booster_.trees_to_dataframe().groupby("tree_index")["node_depth"].max().value_counts().index.max()
+        <= max_depth
+    ), f"Trained model contains trees deeper than max_depth = {max_depth}"
```

Comment:
As currently written, this test could pass even if the `max_depth` implementation wasn't correct. For example, the model you trained could have all empty trees if 0 splits were found.

A much stronger test would look like this:

* train an initial model with `max_depth=7` or something (using parameters to ensure a deterministic fit)
* observe the maximum depth across all trees
* train another model on that same data, with all the same parameters BUT setting `max_depth` to a smaller value
* test that the new model's deepest tree did not exceed `max_depth`